packages/context/src/resolver.ts

```diff
@@ -72,14 +275,44 @@ export function instantiateClass<T>(
  * Resolve the value or promise for a given injection
  * @param ctx Context
  * @param injection Descriptor of the injection
+ * @param session Optional session for binding and dependency resolution
  */
-function resolve<T>(ctx: Context, injection: Injection): ValueOrPromise<T> {
+function resolve<T>(
+  ctx: Context,
+  injection: Injection,
+  session?: ResolutionSession,
+): ValueOrPromise<T> {
+  /* istanbul ignore if */
+  if (debug.enabled) {
+    debug(
+      'Resolving an injection:',
+      ResolutionSession.describeInjection(injection),
+    );
+  }
+  let resolved: ValueOrPromise<T>;
+  // Push the injection onto the session
+  session = ResolutionSession.enterInjection(injection, session);
   if (injection.resolve) {
     // A custom resolve function is provided
-    return injection.resolve(ctx, injection);
+    resolved = injection.resolve(ctx, injection, session);
+  } else {
+    // Default to resolve the value from the context by binding key
+    resolved = ctx.getValueOrPromise(injection.bindingKey, session);
   }
-  // Default to resolve the value from the context by binding key
-  return ctx.getValueOrPromise(injection.bindingKey);
+  if (isPromise(resolved)) {
+    resolved = resolved
+      .then(r => {
+        session!.exitInjection();
+        return r;
+      })
+      .catch(e => {
```

Comment:
I think you should be using a two-callback version of `then` here. Right now, if `exitInjection` throws an error, then this catch handler is invoked for that error, and `exitInjection` is called again.

```ts
resolved.then(
  r => { /*... */ },
  err => { /*... */});
```